### ISSUE TITLE
Async queue

### ISSUE THREAD

**Issue by oroce** [mqttjs/MQTT.js#343]
Hey,

I have a strange setup, my raspberry pi connect to the server through mqtt, but connection sometimes gets dropped, that isn't very problematic since mqtt.js is capable of queueing up the messages.

Sometimes the connection can be back in 1 or 2 minutes but there's a change when there's no connection for 10 minutes but in the meantime watchdog comes in and reboot the device since there was no internet for 5 minutes. And here comes the problem: `mqtt.js` only queues the events in memory  but cannot persist those events to the disk.

Is there somebody who is facing with the same problem? Or do you have recommendation for this case?

First I thought `mqtt-level-store` is for this specific use case but then realized it's more for handling backpressure of messages on the client.
Then I did some experimenting here: https://github.com/oroce/mqttfun

Basically I implemented async queue option, where I needed to change two things:
- ability to pass custom `queue` to the constructor/connect method (there's already a PR for that: https://github.com/mqttjs/MQTT.js/pull/341)
- make `queue.shift` async: https://github.com/oroce/MQTT.js/compare/master...oroce:async-queue

And then I wrote 3 queue adapters:
- memory: https://github.com/oroce/mqttfun/blob/master/queues/memory.js (basically async `[].push` and `[].shift`)
- redis: https://github.com/oroce/mqttfun/blob/master/queues/redis.js
- leveldb: https://github.com/oroce/mqttfun/blob/master/queues/level.js

What do you think? Should I solve this outside of `mqtt.js`, or should be async queue a thing?

**Comment by mcollina** [comment]
You are wrong about the Store API. Its purpose is to store both incoming and outgoing messages and confirmations.

For outgoing, see https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L46-L53.

**Comment by oroce** [comment]
@mcollina: That's okay, my problem happens when `this.connection !== true`: https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L586-L588

**Comment by mcollina** [comment]
If there is no connection, then the packet goes in the queue. There might be a bug in this code, but it was written to handle your use case.

**Comment by oroce** [comment]
Oh no, that does everything as it is intended to do. My problem is that I want a queue which can survive the restart of the client (basically I'm restarting the device itself). For this I need a NON inmemory queue.

**Comment by mcollina** [comment]
mqtt-level-store should be on disk.

**Comment by oroce** [comment]
Yes mqtt-level-store is on disk if the client is online, but if client goes offline the mqtt-level-store wont even receive the message because the message goes into the (in memory) queue.

**Comment by mcollina** [comment]
Before it's connected? Oh, just saw that: it's a bug. Would you mind sending a PR to fix it? https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L586-L588

**Comment by oroce** [comment]
Sure I started working on it, somehow some messages (which were stored during the previous run) arent getting the `puback`s, so I need to run a few more tests.

**Comment by mcollina** [comment]
@oroce any update on this?

**Comment by oroce** [comment]
@mcollina sorry my last month was rather busy but it's still on my list since I want to change [godot](https://github.com/nodejitsu/godot)'s tcp connection to mqtt based connection in my project: https://github.com/oroce/sensortag-with-godot

**Comment by mcollina** [comment]
It might have been fixed. Feel free to reopen if not.
